Gamelogic

JM-Lemmi opened the pull request:
Closes TTTK-7 TTTK-8 TTTK-9

JM-Lemmi committed:
chore(game): add class structure

JM-Lemmi committed:
add game logic

TTTK-7 TTTK-8 TTTK-9

Petzys committed:
fix(game): Typos, Types

Petzys commented:
Most of these classes and methods can be easily tested. Please add tests to verify the functionality.

Petzys commented:
Additionally, we need more documentation in the code. 
At least add doc strings to the classes like [here](https://peps.python.org/pep-0257/).

Petzys commented on Game/rulebase.py:
```diff
@@ -0,0 +1,46 @@
+    def check_win(self, state: GameState) -> int:
+        return self._check_win_horizontal(state) or self._check_win_vertical(state) or self._check_win_diagonal(state)
+    
+    def _check_win_horizontal(self, state: GameState) -> int:
+        for row in state.playfield:
+            if row[0] == row[1] == row[2] and row[0] != 0:
+                return row[0]
+        return 0
+    
+    def _check_win_vertical(self, state: GameState) -> int:
+        for i in range(3):
+            if state.playfield[0][i] == state.playfield[1][i] == state.playfield[2][i] and state.playfield[0][i] != 0:
+                return state.playfield[0][i]
+        return 0
+    
+    def _check_win_diagonal(self, state: GameState) -> int:
+        if state.playfield[0][0] == state.playfield[1][1] == state.playfield[2][2] and state.playfield[0][0] != 0:
+            return state.playfield[0][0]
+        elif state.playfield[0][2] == state.playfield[1][1] == state.playfield[2][0] and state.playfield[0][2] != 0:
+            return state.playfield[0][0]
+        return 0        
```
These methods are not capable of handling playfields larger than 3*3. 
Additionally, this just looks straight up ugly. Please try to use iteration over the playfield dimensions etc.

JM-Lemmi replied:
isnt that the point of this class? Any other ruleset should be an extension of this class with overwritten functions.

JM-Lemmi replied:
c2613fa

Petzys commented on Game/rulebase.py:
```diff
@@ -0,0 +1,46 @@
+        return True #TODO
+    
+    def explain(self, state: GameState, prev: (int, int), new: (int, int)) -> str:
+        return "" #TODO
```
I just realized that it makes no sense at all to have a "prev" parameter for explain() since TicTacToe does not work that way. I already changed the class diagram. Please remove the "prev" parameter. We only need "new" and the state.

Additionally, Implement these functions. 
Explain can be very generic (e.g. "This field is already taken. Take an empty one". "It is not your turn").
Remember to also implement the OK case: Explain can be called on legal moves too.

Petzys commented on Game/game.py:
```diff
@@ -0,0 +1,23 @@
+
+class Game:
+    def __init__(self, player1: Player, player2: Player, rule_base: RuleBase = RuleBase()):
+        self._id: int = 1 #TODO count up
```
Using a random number with a high entropy should be enough in this case. (Or do something else)

Petzys replied:
TODO

Petzys commented on Game/rulebase.py:
```diff
@@ -0,0 +1,46 @@
+    
+    # returns the winning player id
+    def check_win(self, state: GameState) -> int:
+        return self._check_win_horizontal(state) or self._check_win_vertical(state) or self._check_win_diagonal(state)
```
This method does not implement a draw when all field are filled but none of the players won.

I recommend to change the GameStates winner and finish attribute here. Afterwards, Game checks for the attributes of game states and evaluates whether it is draw or who is the winner

JM-Lemmi replied:
43655aa

Petzys commented on Game/game.py:
```diff
@@ -0,0 +1,23 @@
+        
+        self.state.set_player_position(player, new_position)
+
+        if winner := (self.rule_base.check_win(self.state) != 0):
+            return self.players[winner]
+        else:
+            return None
```
Also think about the draw possibility here

JM-Lemmi replied:
43655aa

Petzys commented on Game/rulebase.py:
```diff
@@ -0,0 +1,46 @@
+    def playfield_dimensions(self) -> (int, int):
+        return self._playfield_dimensions
+
+    def is_move_valid(self, state: GameState, new_position: (int, int)) -> bool:
+        if state.playfield_value(new_position) == 0:
+            return True
+        elif state.playfield_value(new_position) == 1 or state.playfield_value(new_position) == 2:
+            return False
+        else:
+            raise ValueError("Invalid playfield value")
```
You should also check if the player who wants to move is the current player.

JM-Lemmi replied:
Mit den gegebenen werten kann diese Funktion das gar nicht checken.

JM-Lemmi replied:
Wird außerdem von is_state_valid abgedeckt falls man das danach checken möchte. ich wäre einfach dafür, dass die usptream funktionen tracken, wer grade dran ist.

Petzys commented on Game/game.py:
```diff
@@ -0,0 +1,23 @@
+    # throws illegal move error on illegal move
+    def move(self, player: int, new_position: (int, int)) -> Player | None:
+        if not self.rule_base.is_move_valid(self.state, new_position):
+            raise ValueError("Illegal move")
```
In this line you should Raise the output of RuleBase.explain(). 

Additionally, I would recommend using a custom Error Class, that you can raise. This allows for easier storage of error related values e.g. the position which is not allowed. But this is optional.

JM-Lemmi replied:
I hate the explain function anyway. Ist das eine muss-anforderung? sonst würde ich das streichen.

Petzys replied:
Actually yes. This is necessary.

Petzys replied:
TODO

JM-Lemmi committed:
fix: handle playfields larger than 3x3 in check_win
refractor check_win classes into one

JM-Lemmi committed:
docs: add doc strings to all classes and function

JM-Lemmi committed:
fix: align List indices with player id

JM-Lemmi committed:
feat(game): add draw
move no longer returns the winning player, calling functions will need to check game state themselves

JM-Lemmi committed:
feat(game): implement is_game_state_valid

JM-Lemmi commented:
Ich würde auch gerne den `GameState.current_player()` entfernen, das ist einfach nicht nötig.

JM-Lemmi committed:
remove `current_player` from GameState

JM-Lemmi committed:
doc: fix typing of tuples and 2dimension-arrays

JM-Lemmi committed:
test(game): add tests to gamestate

JM-Lemmi committed:
test: add tests for rulebase

JM-Lemmi committed:
ci: run tests with workflow

Petzys commented on Game/rulebase.py:
```diff
@@ -0,0 +1,141 @@
+            str: The explanation for the move.
+
+        """
+        return ""  # TODO
```
Please implement

JM-Lemmi replied:
Wie gesagt würde ich das gerne eigentlich streichen, außer es gibt dazu spezifische Anforderungen

JM-Lemmi replied:
Ah hab den Kommentar gefunden.

Petzys commented on Game/test_gamestate.py:
For next time: Please use python unittests. Do not implement tests yourself using basic asserts. But its ok for now.

Petzys reviewed (changes requested):
Well done and thanks! Only a few things still left to do but this is very nice!

Petzys committed:
Feat(gamelogic): Moved explain() inside is_move_valid

Petzys committed:
Fix(gamelogic): Draw check did not work

Petzys committed:
Refactor(gamelogic tests): Now using python unittest instead of basic asserts

Petzys committed:
Feat(gamelogic): TTTK-50 Improved explain functionality

Petzys committed:
Feat(gamelogic): Added UUID for Game and Player Classes TTTK-7 TTTK-8 TTTK-9
